src/lib/layout/Searchbar.tsx

```diff
@@ -7,126 +7,209 @@ import {
         >
           {value}
         </Text>
       )}
-    </ListItem>
+    </StyledListItem>
   );
 };
 
-// TODO - Implement all search for Wasm chain
-const Searchbar = () => {
-  const navigate = useInternalNavigate();
-  const { validateContractAddress, validateUserAddress } = useValidateAddress();
+const ResultRender = ({
+  results,
+  keyword,
+  cursor,
+  setCursor,
+  handleSelectResult,
+}: {
+  results: SearchResultType[];
+  keyword: string;
+  cursor: Option<number>;
+  setCursor: (index: Option<number>) => void;
+  handleSelectResult: (type?: SearchResultType, isClick?: boolean) => void;
+}) => (
+  <>
+    {!results.length ? (
+      <StyledListItem>
+        <Text variant="body2" fontWeight={500} color="text.dark" p="8px">
+          {NOT_FOUND}
+        </Text>
+      </StyledListItem>
+    ) : (
+      results.map((type, index) => (
+        <ResultItem
+          index={index}
+          key={type}
+          type={type}
+          value={keyword}
+          cursor={cursor}
+          setCursor={setCursor}
+          handleSelectResult={handleSelectResult}
+        />
+      ))
+    )}
+  </>
+);
 
+const getNextCursor = (
+  key: string,
+  current: Option<number>,
+  lastIndex: number
+) => {
+  switch (key) {
+    case "ArrowUp":
+      if (current === undefined) return lastIndex;
+      return current <= 0 ? lastIndex : current - 1;
+    case "ArrowDown":
+      if (current === undefined) return 0;
+      return current >= lastIndex ? 0 : current + 1;
+    default:
+      return undefined;
+  }
+};
+
+const Searchbar = () => {
   const [keyword, setKeyword] = useState("");
   const [displayResults, setDisplayResults] = useState(false);
-  const [results, setResults] = useState<SearchResultType[]>([]);
+  const [isTyping, setIsTyping] = useState(false);
+  const [cursor, setCursor] = useState<number>();
 
+  const navigate = useInternalNavigate();
+  const resetHandlerStates = useCallback(() => {
+    setIsTyping(false);
+    setCursor(undefined);
+  }, []);
+  const { results, isLoading } = useSearchHandler(keyword, resetHandlerStates);
   const boxRef = useRef<HTMLDivElement>(null);
-  const chainConfig = getChainConfig();
-
-  let placeholder = "Search by Wallet Address / Tx Hash / ";
-  placeholder += chainConfig.isWasm ? "Code ID / Contract Address" : "Block";
-
-  useEffect(() => {
-    const res: SearchResultType[] = [];
-    if (chainConfig.isWasm && isCodeId(keyword)) res.push("Code ID");
-    else if (isBlock(keyword)) res.push("Block");
-
-    if (!validateContractAddress(keyword)) res.push("Contract Address");
-    if (!validateUserAddress(keyword)) res.push("Wallet Address");
-    if (isTxHash(keyword)) res.push("Transaction Hash");
-    // TODO: add proposal ID
-    setResults(res);
-  }, [
-    chainConfig.isWasm,
-    keyword,
-    validateContractAddress,
-    validateUserAddress,
-  ]);
 
   const handleSearchChange = (e: ChangeEvent<HTMLInputElement>) => {
```

Comment:
Should the first row of the item list be highlighted automatically when some results show in the list?